### PR TITLE
Problem: missing IBCChannelMessage projection

### LINKED ISSUE #520: Problem: missing IBCChannelMessage projection

### PR DESCRIPTION
**Pull request by ysong42**
Solution: fixed #520 

## Changes

- Introduced new projection `IBCChannelMessage`.
- Introduced new table for the projection:
```
CREATE TABLE view_ibc_channel_messages (
    id BIGSERIAL,
    channel_id VARCHAR NOT NULL,
    block_height BIGINT NOT NULL,
    block_time BIGINT NOT NULL,
    transaction_hash VARCHAR NOT NULL,
    success BOOLEAN NOT NULL,
    error VARCHAR NOT NULL,
    sender VARCHAR NOT NULL,
    receiver VARCHAR NOT NULL,
    denom VARCHAR NOT NULL,
    amount VARCHAR NOT NULL,
    message_type VARCHAR NOT NULL,
    message JSONB NOT NULL,
    PRIMARY KEY (id)
);

CREATE INDEX view_ibc_channel_messages_channel_id_btree_index ON view_ibc_channel_messages USING btree (channel_id);
```
- Introduced new API `/ibc/channels/{channelId}/messages`

## About the new API `/ibc/channels/{channelId}/messages`

By default, messages sorted by blockTime in descending order.
User could set `order=blockTime.asc` to sort it in ascending order.

User could also specify a filter of the returned messageType, for example `filter.msgType=MsgTransfer,MsgAcknowledgement`.

## Sample API Response

```
{
    "result": [
        {
            "channelId": "channel-1",
            "blockHeight": 69165,
            "blockTime": "2021-08-23T11:41:13.9824503Z",
            "transactionHash": "BFEE66D3D8F00A2841B41A726A31FB8455AB5D5E8055F5418FE87BC8ABB5CE9B",
            "success": true,
            "error": "",
            "sender": "",
            "receiver": "",
            "denom": "",
            "amount": "",
            "messageType": "MsgChannelOpenAck",
            "message": {
                "name": "MsgChannelOpenAckCreated",
                "uuid": "6b0c77ec-cb31-4fae-8f00-c7e03f17a5ca",
                "height": 69165,
                "params": {
                    "counterpartyChannelId": "channel-lgfU",
                    "channelId": "channel-1",
                    "proofHeight": {
                        "revisionHeight": "4",
                        "revisionNumber": "0"
                    },
                    "counterpartyPortId": "transfer",
                    "counterpartyVersion": "ics20-1",
                    "portId": "transfer",
                    "signer": "tcro1xkm3243sdf35dm6evrlh8u9a6faj28m357yva5",
                    "proofTry": "CkQKQhJAaMS8fKYj9FAxS47yjPeG9TcHK1eW9kQ00PkMzVLBtzcOg9B/AuStHv6pEYz/htY00oyOfbNm2jNHRI5P5LB9BRCsko6JBg==",
                    "connectionId": "connection-1"
                },
                "txHash": "BFEE66D3D8F00A2841B41A726A31FB8455AB5D5E8055F5418FE87BC8ABB5CE9B",
                "msgName": "MsgChannelOpenAck",
                "version": 1,
                "msgIndex": 0
            }
        },
        {
            "channelId": "channel-1",
            "blockHeight": 69164,
            "blockTime": "2021-08-23T11:41:08.574463984Z",
            "transactionHash": "B4A25CCA3D5A09BC3423CBF4710E73C0B0313F911DE6CEB7A5700342F73B7D07",
            "success": true,
            "error": "",
            "sender": "",
            "receiver": "",
            "denom": "",
            "amount": "",
            "messageType": "MsgChannelOpenInit",
            "message": {
                "txHash": "B4A25CCA3D5A09BC3423CBF4710E73C0B0313F911DE6CEB7A5700342F73B7D07",
                "msgName": "MsgChannelOpenInit",
                "version": 1,
                "msgIndex": 0,
                "name": "MsgChannelOpenInitCreated",
                "uuid": "33b51f70-2558-4820-b963-65810568154e",
                "height": 69164,
                "params": {
                    "channelId": "channel-1",
                    "portId": "transfer",
                    "signer": "tcro1xkm3243sdf35dm6evrlh8u9a6faj28m357yva5",
                    "channel": {
                        "state": "STATE_INIT",
                        "version": "ics20-1",
                        "ordering": "ORDER_UNORDERED",
                        "counterparty": {
                            "portId": "transfer",
                            "channelId": ""
                        },
                        "connectionHops": [
                            "connection-1"
                        ]
                    }
                }
            }
        }
    ],
    "pagination": {
        "total_record": 2,
        "total_page": 1,
        "current_page": 1,
        "limit": 20
    }
}
```

### CODE REVIEW

**Comment by ysong42** [comment]
This is the continue of PR: https://github.com/crypto-com/chain-indexing/pull/521

As the old PR files diff is messed up. Previous comments can be traced on the old PR.

File: projection/ibc_channel_message/view/ibc_channel_messages_total.go [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+	"github.com/crypto-com/chain-indexing/appinterface/rdb"
+)
+
+type IBCChannelMessagesTotal struct {
+	*view.Total
+}
+
+func NewIBCChannelMessagesTotal(rdbHandle *rdb.Handle) *IBCChannelMessagesTotal {
+	return &IBCChannelMessagesTotal{
+		view.NewTotal(rdbHandle, "view_ibc_channel_messages_total"),
```
Comment (davcrypto) [inline projection/ibc_channel_message/view/ibc_channel_messages_total.go:14]: Do we have an endpoint to get this data?

Comment (ysong42) [reply projection/ibc_channel_message/view/ibc_channel_messages_total.go:14]: Nope, It is mainly for pagination purpose. Not for any endpoint to direct query it.

Shamelessly copied from other projection 😂 . For example, `accountTransactions` has a similar view...

File: migrations/20210910114750_view_ibc_channel_messages.up.sql [hunk 2]
```diff
@@ -0,0 +1,17 @@
+CREATE TABLE view_ibc_channel_messages (
+    id BIGSERIAL,
+    channel_id VARCHAR NOT NULL,
```
Comment (davcrypto) [inline migrations/20210910114750_view_ibc_channel_messages.up.sql:3]: index

File: projection/ibc_channel_message/view/ibc_channel_messages_total.go (hunk 1 above)
Comment (davcrypto) [reply projection/ibc_channel_message/view/ibc_channel_messages_total.go:14]: Thanks, got it!

**Comment by davcrypto** [comment]
Thanks @ysong42 for the new projection! Only left one more comment about the database index, other than that LGTM!

File: migrations/20210910114750_view_ibc_channel_messages.up.sql (hunk 2 above)
Comment (ysong42) [reply migrations/20210910114750_view_ibc_channel_messages.up.sql:3]: Good catch! Thanks!

**Review by calvinaco — COMMENTED** [review]
Overall looks good to me, left a comment on DB schema technical question.

File: migrations/20210910114750_view_ibc_channel_messages.up.sql [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+    transaction_hash VARCHAR NOT NULL,
+    relayer VARCHAR NOT NULL,
+    success BOOLEAN NOT NULL,
+    error VARCHAR NOT NULL,
+    sender VARCHAR NOT NULL,
+    receiver VARCHAR NOT NULL,
+    denom VARCHAR NOT NULL,
+    amount VARCHAR NOT NULL,
+    message_type VARCHAR NOT NULL,
+    message JSONB NOT NULL,
```
[the remark this answers is not in the document]
Comment (ysong42) [reply migrations/20210910114750_view_ibc_channel_messages.up.sql:15]: Good catch! Thanks! Let me update it!

**Comment by ysong42** [comment]
Thanks @calvinaco @davcrypto for the reviews! The code is updated! Might need another round review 🙏

**Review by calvinaco — APPROVED** [review]

**Review by davcrypto — APPROVED** [review]

**Pull request merged**